### PR TITLE
New TermInterface

### PR DESCRIPTION
**Pull request by 0x0f0f0f**

### CODE REVIEW

**Comment by shashi** [comment]
The way I implemented `children` is by making it return the operation and arguments together. The `head` for BasicSymbolics was `basicsymbolic`. See: https://github.com/JuliaSymbolics/SymbolicUtils.jl/compare/s/terminterface1?expand=1

**Comment by shashi** [comment]
@0x0f0f0f does that make sense? I wanna finish this up and try out MT on SU expressions

**Comment by 0x0f0f0f** [comment]
> @0x0f0f0f does that make sense? I wanna finish this up and try out MT on SU expressions

This is up to you, if you think it makes more sense to have it implemented as such, please go ahead.

**Comment by 0x0f0f0f** [comment]
@shashi adjusted with your proposal - CI passing now. documentation job is broken. disabled fuzzing tests as they segfault until https://github.com/JuliaMath/SpecialFunctions.jl/issues/446 is fixed

File: src/types.jl [hunk 1]
```diff
@@ -545,9 +548,11 @@ different type than `t`, because `f` also influences the result.
   resulting similar term to this type.
 """
 similarterm(t::Symbolic, f, args; metadata=nothing) =
-    similarterm(t, f, args, _promote_symtype(f, args); metadata=metadata)
-similarterm(t::BasicSymbolic, f, args,
-            symtype; metadata=nothing) = basic_similarterm(t, f, args, symtype; metadata=metadata)
+    maketerm(typeof(t), f, args, _promote_symtype(f, args); metadata)
+similarterm(t::BasicSymbolic, f, args, symtype; metadata=nothing) =
+    maketerm(typeof(t), f, args, symtype; metadata=metadata)
+maketerm(T::Type{<:Symbolic}, f, args, symtype; metadata=nothing) =
```
Comment (shashi) [inline src/types.jl:554]: Arrays are also <: Symbolic so maybe just make it BasicSymbolic?

**Comment by 0x0f0f0f** [comment]
Downstream MTK Failing due to deprecation warnings
```
 Discrete System: Test Failed at /cache/build/builder-amdci5-1/julialang/julia-release-1-dot-10/usr/share/julia/stdlib/v1.10/Test/src/Test.jl:903
  Expression: isempty(stderr_content)
   Evaluated: isempty("WARNING: ModelingToolkit.istree is deprecated, use TermInterface.iscall instead.\n  likely near /home/runner/work/SymbolicUtils.jl/SymbolicUtils.jl/downstream/test/discrete_system.jl:267\nin collect_var_to_name! at /home/runner/work/SymbolicUtils.jl/SymbolicUtils.jl/downstream/src/utils.jl\n")
``` 

I suggested to remove deprecation warnings and do a major release bump

**Comment by 0x0f0f0f** [comment]
> I suggested to remove deprecation warnings and do a major release bump

Ok, there was already a major version bump, shouldn't downstream CI not be executed?

**Comment by 0x0f0f0f** [comment]
@shashi any news on the state of this?

**Comment by shashi** [comment]
I think we can release it as a major release and update downstream. The errors are all because of the depwarn. Some packages are testing the if stderr is empty.

**Comment by shashi** [comment]
I'll make the relevant PRs to Symbolics and MTK today.

**Comment by devmotion** [comment]
Shouldn't this have been a breaking major release? It completely broke my project (that due to Catalyst still has to use MTK v8) when running `Pkg.update()`.

File: docs/src/index.md [hunk 2]
```diff
@@ -108,15 +108,8 @@ g(2//5, g(1, β))
...
-All the expression types support the following:
-
-- `istree(x)` -- always returns `true` denoting, `x` is not a leaf node like Sym or a literal.
-- `operation(x)` -- the function being called
-- `arguments(x)` -- a vector of arguments
-- `symtype(x)` -- the "inferred" type (`T`)
-
-See more on the interface [here](/interface)
-
+All the expression types support the [TermInterface.jl](https://github.com/0x0f0f0f/TermInterface.jl) interface.
```
Comment (bowenszhu) [inline docs/src/index.md:111]: The url should be https://github.com/JuliaSymbolics/TermInterface.jl

**Comment by 0x0f0f0f** [comment]
> Shouldn't this have been a breaking major release? It completely broke my project (that due to Catalyst still has to use MTK v8) when running `Pkg.update()`.

@shashi I think this should have been a major breaking release. It has also been reported by a colleague of mine. Lots of deprecation warnings. I think we should just do a major release and drop `similarterm` alltogether and use maketerm

**Pull request merged**